nova_vm/src/ecmascript/builtins/typed_array/data.rs

```diff
@@ -3,13 +3,95 @@
 // file, You can obtain one at https://mozilla.org/MPL/2.0/.
 
 use crate::{
-    ecmascript::types::OrdinaryObject,
+    ecmascript::{
+        builtins::{
+            array_buffer::{ViewedArrayBufferByteLength, ViewedArrayBufferByteOffset},
+            ArrayBuffer,
+        },
+        types::OrdinaryObject,
+    },
     heap::{CompactionLists, HeapMarkAndSweep, WorkQueues},
 };
 
-#[derive(Debug, Clone, Default)]
+#[repr(transparent)]
+#[derive(Debug, Clone, Copy, Eq, PartialEq)]
+pub(crate) struct TypedArrayArrayLength(pub u32);
+
+impl TypedArrayArrayLength {
+    pub const fn value(value: u32) -> Self {
+        Self(value)
+    }
+
+    /// A sentinel value of `u32::MAX - 1` means that the byte length is stored in an
+    /// associated map in the heap. This will most likely be a very rare case,
+    /// only applicable for 4GB+ buffers.
+    pub const fn heap() -> Self {
+        Self(u32::MAX - 1)
+    }
+
+    /// A sentinel value of `u32::MAX` means that the byte length is the
+    /// `AUTO` value used in the spec.
+    pub const fn auto() -> Self {
+        Self(u32::MAX)
+    }
+}
+
+impl Default for TypedArrayArrayLength {
+    fn default() -> Self {
+        Self::auto()
+    }
+}
+
+impl From<Option<usize>> for TypedArrayArrayLength {
+    fn from(value: Option<usize>) -> Self {
+        match value {
+            Some(value) => {
+                if value >= Self::heap().0 as usize {
+                    Self::heap()
+                } else {
+                    Self::value(value as u32)
+                }
+            }
+            None => Self::auto(),
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
 pub struct TypedArrayHeapData {
     pub(crate) object_index: Option<OrdinaryObject>,
+    /// ### [\[\[ViewedArrayBuffer\]\]](https://tc39.es/ecma262/#sec-properties-of-typedarray-instances)
+    pub(crate) viewed_array_buffer: ArrayBuffer,
+    /// ### [\[\[ByteLength\]\]](https://tc39.es/ecma262/#sec-properties-of-typedarray-instances)
+    pub(crate) byte_length: ViewedArrayBufferByteLength,
+    /// ### [\[\[ByteOffset\]\]](https://tc39.es/ecma262/#sec-properties-of-typedarray-instances)
+    pub(crate) byte_offset: ViewedArrayBufferByteOffset,
+    /// ### [\[\[ArrayLength\]\]](https://tc39.es/ecma262/#sec-properties-of-typedarray-instances)
+    pub(crate) array_length: TypedArrayArrayLength,
```

Comment:
**thought:** We don't actually need both the byte length and the array length, since `byte_length = 4 * array_length` must always hold true. That being said... With that we wouldn't have information about the type of TA this heap data belongs to. As long as we always check the byte length then this shouldn't really matter but if we eg. decide to save the array length and calculate byte length based on that then now we run into a safety issue where an attacker could take a Uint8Array and change its variant to BigUint64Array, thus enabling reading and writing past the ArrayBuffer end in memory.

Well, we still do not have information about that so an attacker could take eg. a Uint16Array and turn that into a Float16Array or something like that. But that's hardly dangerous.